general/okina.hpp

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#ifndef MFEM_OKINA_HPP
+#define MFEM_OKINA_HPP
+
+// *****************************************************************************
+#include "../config/config.hpp"
+#include "../general/error.hpp"
+
+// *****************************************************************************
+#include <list>
+#include <cmath>
+#include <cassert>
+#include <cstring>
+#include <iostream>
+#include <unordered_map>
+
+// *****************************************************************************
+#include "./cuda.hpp"
+#include "./occa.hpp"
+
+// *****************************************************************************
+#include "mm.hpp"
+#include "config.hpp"
+
+// *****************************************************************************
+// * GPU & HOST FOR_LOOP bodies wrapper
+// *****************************************************************************
+template <size_t BLOCK_SZ, typename DBODY, typename HBODY>
+void wrap(const size_t N, DBODY &&d_body, HBODY &&h_body)
+{
+   const bool gpu = mfem::config::usingGpu();
+   if (gpu)
+   {
+      return cuWrap<BLOCK_SZ>(N,d_body);
+   }
+   for (size_t k=0; k<N; k+=1) { h_body(k); }
+}
+
+// *****************************************************************************
+// * MFEM_FORALL splitter
+// *****************************************************************************
+#define MFEM_FORALL(i, N, B)                                            \
```

Comment:
I was wondering if there might be situations where we might need the iteration space to not start at 0? Here we assume the range is [0, N), but could there be a need for [a, N)? where a > 0.